File: libs/east/src/location.ts
/**
 * Copyright (c) 2025 Elara AI Pty Ltd
 * Dual-licensed under AGPL-3.0 and commercial license. See LICENSE for details.
 *
 * @remarks
 */
export type Location = {
  /** The source file path */
  filename: string,
  /** The 1-based line number */
  line: bigint,
  /** The 1-based column number */
  column: bigint,
}

/**
 * Formats a single location as a human-readable string.
 *
 * @param location - The source location to format
 * @returns A string in the format `"<filename> <line>:<column>"`
 *
 * @example
 * ```ts
 * const loc = { filename: "main.ts", line: 42n, column: 15n };
 * printLocation(loc); // "main.ts 42:15"
 * ```
 */
export function printLocation(location: Location): string {
  return `${location.filename} ${location.line}:${location.column}`;
}

/**
 * Formats an array of locations as a stack trace string.
 *
 * @param locations - The source locations to format
 * @returns A stack trace string with each location on a new line
 *
 * @example
 * ```ts
 * const locs = [
 *   { filename: "main.ts", line: 42n, column: 15n },
 *   { filename: "lib.ts", line: 10n, column: 5n }
 * ];
 * printLocations(locs);
 * // "main.ts 42:15
 * //   at lib.ts 10:5"
 * ```
 */
export function printLocations(locations: Location[]): string {
  if (locations.length === 0) return '<unknown>';
  const [first, ...rest] = locations;
  const header = printLocation(first!);
  if (rest.length === 0) return header;
  return header + '\n' + rest.map(loc => `  at ${printLocation(loc)}`).join('\n');
}

/**
 * Determines if a stack frame should be included in location captures.
 * Filters out internal frames (Node.js internals, node_modules) to show
 * only user code in error stack traces.
 *
 * @param filename - The file path from the stack frame
 * @returns true if the frame should be included, false to filter it out
 */
/**
 * Directories holding East's own modules, given the URL of this module.
 *
 * Both the compiled and the original tree are returned, because `stack`
 * reports the compiled path normally and the original one under
 * `--enable-source-maps` — a frame is East's either way.
 *
 * @param moduleUrl - `import.meta.url` of this module, or undefined where the
 *   bundler does not provide one (a CJS or browser bundle, which has no East
 *   tree to point at)
 * @returns Directory prefixes to treat as East's own, empty when unknowable
 *
 * @remarks
 * Exported for testing. An empty result means frame filtering falls back to
 * the `node_modules` rule, which is the correct answer for a bundle: East's
 * code shares a file with its caller there, so no path can separate them.
 */
export function eastOwnDirs(moduleUrl: string | undefined): readonly string[] {
  if (moduleUrl === undefined || moduleUrl === '') return [];
  const path = normalizeSeparators(stripFileUrl(moduleUrl));
  const slash = path.lastIndexOf('/');
  if (slash === -1) return [];
  const dir = path.slice(0, slash);
  const root = dir.endsWith('/dist/src') ? dir.slice(0, -'/dist/src'.length)
    : dir.endsWith('/src') ? dir.slice(0, -'/src'.length)
      : null;
  return root === null ? [`${dir}/`] : [`${root}/src/`, `${root}/dist/src/`];
}

/** East's own module directories in this environment (see {@link eastOwnDirs}). */
const EAST_OWN_DIRS: readonly string[] = eastOwnDirs(import.meta.url as string | undefined);

function shouldIncludeFrame(filename: string): boolean {
  // Skip Node.js internal modules (e.g., node:internal/modules/...)
  if (filename.startsWith('node:')) return false;

  // Skip node_modules - filters out third-party packages including East
  // when installed as a dependency. Handles both Unix and Windows paths.
  if (filename.includes('/node_modules/') || filename.includes('\\node_modules\\')) return false;

  // Skip East's own modules however they were loaded. The node_modules rule
  // covers an installed East but not a linked or in-repo one, which would
  // otherwise head every trace with East internals instead of user code — and
  // bake East's own line numbers into serialized source maps. East's specs sit
  // in the same tree but are callers of the library, not part of it.
  const path = normalizeSeparators(stripFileUrl(filename));
  if (!/\.spec\.[cm]?[jt]s$/.test(path) && EAST_OWN_DIRS.some(dir => path.startsWith(dir))) return false;

  return true;
}

/**
 * Captures the call stack as an array of source locations, filtered to
 * show only user code.
 *
 * @returns An array of {@link Location} objects representing the call stack,
 *          from innermost (most recent) to outermost frame
 *
 * @remarks
 * This function uses JavaScript's Error stack traces to capture the call
 * stack. Internal frames (Node.js internals, node_modules packages) are
 * filtered out to provide clean stack traces showing only user code.
 * Returns an empty array if the stack trace cannot be parsed.
 *
 * @example
 * ```ts
 * function myFunction() {
 *   const stack = get_location(); // Gets call stack of user code
 *   console.log(printLocations(stack));
 * }
 * ```
 */


// ── Frame path normalization ───────────────────────────────────────────
//
// Stack-trace paths are absolute and environment-specific (and `file://` URLs
// under ESM). Baking them into serialized IR leaks the author's filesystem
// layout and makes the IR non-reproducible across machines and build modes. We
// normalize each frame path to a portable form: strip the `file://` scheme,
// normalize separators, and relativize to a base directory. The base defaults
// to the working directory when one is available (Node) and can be set with
// `setLocationBasePath` (e.g. for deterministic fixtures). Browser / no-cwd
// environments keep the cleaned path. This stays free of `node:` imports
// (browser-safe) and never throws.

let explicitBaseSet = false;
let explicitBase: string | undefined;
let autoBase: string | undefined;
let autoBaseComputed = false;

function normalizeSeparators(p: string): string {
  return p.replace(/\\/g, '/');
}

function stripTrailingSlash(p: string): string {
  return p.length > 1 && p.endsWith('/') ? p.slice(0, -1) : p;
}

function stripFileUrl(p: string): string {
  if (!p.startsWith('file://')) return p;
  try {
    // `URL` is a universal (WHATWG) global — present in Node and browsers, no
    // `node:url` import. `.pathname` is percent-encoded; decode it.
    const pathname = decodeURIComponent(new URL(p).pathname);
    return pathname.replace(/^\/([A-Za-z]:)/, '$1'); // Windows: "/C:/x" → "C:/x"
  } catch {
    return p;
  }
}

function computeAutoBase(): string | undefined {
  if (!autoBaseComputed) {
    autoBaseComputed = true;
    // Guarded global access (not a `node:` import) — `process` is undefined in
    // the browser, where we simply skip relativization.
    const proc = (globalThis as { process?: { cwd?: () => string } }).process;
    if (proc !== undefined && typeof proc.cwd === 'function') {
      try {
        autoBase = stripTrailingSlash(normalizeSeparators(String(proc.cwd())));
      } catch {
        autoBase = undefined;
      }
    }
  }
  return autoBase;
}

function currentBase(): string | undefined {
  return explicitBaseSet ? explicitBase : computeAutoBase();
}

/**
 * Set the base directory that captured locations are relativized against.
 * Pass a path to relativize against it (paths outside it stay absolute); pass
 * `undefined` to reset to the automatic default (the working directory in
 * Node, none in the browser).
 */
export function setLocationBasePath(base: string | undefined): void {
  if (base === undefined) {
    explicitBaseSet = false;
    explicitBase = undefined;
  } else {
    explicitBaseSet = true;
    explicitBase = stripTrailingSlash(normalizeSeparators(stripFileUrl(base)));
  }
}

/**
 * Normalize a stack-frame path to a portable, deterministic form: `file://`
 * stripped, separators as `/`, and relativized to the current base when the
 * path is under it. Exported for testing. Never throws.
 */
export function normalizeFramePath(raw: string): string {
  try {
    const cleaned = normalizeSeparators(stripFileUrl(raw));
    const base = currentBase();
    if (base !== undefined && base !== '') {
      if (cleaned === base) return '.';
      if (cleaned.startsWith(base + '/')) return cleaned.slice(base.length + 1);
    }
    return cleaned;
  } catch {
    return raw;
  }
}

function capture_stack_frames(): Location[] {
  const err = new Error();
  const stack = err.stack;
  if (!stack) return [];

  const lines = stack.split('\n').slice(1); // Skip "Error" line
  const frames: Location[] = [];

  for (const line of lines) {
    // Take the path before :line:col (allows spaces and a Windows drive colon).
    const match = line.match(/\((.+):(\d+):(\d+)\)\s*$/) ?? line.match(/\bat\s+(.+):(\d+):(\d+)\s*$/);
    if (match) {
      const [, filename, lineNum, column] = match;
      // `shouldIncludeFrame` runs on the raw path so its node_modules/internal
      // filters see absolute paths; the stored path is normalized.
      if (filename && filename !== '' && shouldIncludeFrame(filename)) {
        frames.push({ filename: normalizeFramePath(filename), line: BigInt(lineNum!), column: BigInt(column!) });
      }
    }
  }

  return frames;
}

export function get_location(): Location[] {
  return capture_stack_frames();
}

// ── SourceMap ──────────────────────────────────────────────────────────

/** Reserved sentinel: resolve(0n) returns [], the "no/unknown location" id. */
export const UNKNOWN_LOC_ID: bigint = 0n;

export class SourceMap {
  /** Location stacks, indexed by Number(loc_id). stacks[0] is always []. */
  private readonly stacks: Location[][] = [[]];
  /** Content key → loc_id for dedup during construction. */
  private readonly intern = new Map<string, bigint>();

  /** Intern a location stack. Returns a stable loc_id (bigint).
   *  Equal content returns the same id. Empty stack → UNKNOWN_LOC_ID. */
  intern_stack(stack: Location[]): bigint {
    if (stack.length === 0) return UNKNOWN_LOC_ID;
    const key = stack.map(l => `${l.filename}|${l.line}|${l.column}`).join('\n');
    const existing = this.intern.get(key);
    if (existing !== undefined) return existing;
    const id = BigInt(this.stacks.length);
    this.stacks.push(stack);
    this.intern.set(key, id);
    return id;
  }

  /** Resolve a loc_id to its stack. Returns [] for UNKNOWN_LOC_ID or out-of-range. */
  resolve(loc_id: bigint): readonly Location[] {
    return this.stacks[Number(loc_id)] ?? [];
  }

  /** Total number of entries, including the reserved empty entry at index 0. */
  get size(): bigint { return BigInt(this.stacks.length); }

  /** All entries in loc_id order (for serialization). Entry 0 is always []. */
  entries(): readonly Location[][] { return this.stacks; }
}

// ── Scoped context ─────────────────────────────────────────────────────

let _currentMap: SourceMap | null = null;

/**
 * Matches the `loc_id <n>` that compile-time error messages embed.
 *
 * Errors raised while building or lowering IR name the offending node by its
 * loc_id, because an id is all the IR carries. Any message that does so must
 * use exactly this form so {@link resolveLocIds} can give it meaning.
 */
const LOC_ID_IN_MESSAGE = /\bloc_id (\d+)\b/g;

/**
 * Replace every `loc_id <n>` in an error message with the source location it
 * names.
 *
 * @param message - Error message that may embed loc_ids
 * @param map - Source map to resolve them against
 * @returns The message with each id replaced by `file:line:column`, or by
 *   `an unknown location` where the map has no entry for it
 *
 * @remarks
 * An id the map cannot resolve still names nothing the reader can act on, so
 * it is described rather than printed.
 */
export function resolveLocIds(message: string, map: SourceMap): string {
  return message.replace(LOC_ID_IN_MESSAGE, (_whole, id: string) => {
    const [location] = map.resolve(BigInt(id));
    return location ? `${location.filename}:${location.line}:${location.column}` : 'an unknown location';
  });
}

/** Run `fn` with `map` as the current scope's source map. Re-entrant safe. */
export function with_source_map<T>(map: SourceMap, fn: () => T): T {
  const prev = _currentMap;
  _currentMap = map;
  try {
    return fn();
  } catch (e: unknown) {
    // The map that gives a loc_id meaning is only in scope here, so this is the
    // last point at which an error carrying one can be made readable.
    if (e instanceof Error) e.message = resolveLocIds(e.message, map);
    throw e;
  } finally {
    _currentMap = prev;
  }
}

/** Wraps `fn` in a new `with_source_map` only if no map is currently active.
 *  Nested calls inherit the existing map. */
export function ensure_source_map<T>(fn: () => T): T {
  if (_currentMap) return fn();
  return with_source_map(new SourceMap(), fn);
}

/** Returns the current SourceMap, or null if outside any with_source_map scope. */
export function get_current_source_map(): SourceMap | null {
  return _currentMap;
}

/** Capture current stack and intern into the active SourceMap.
 *  Returns UNKNOWN_LOC_ID (0n) if no map is active. */
export function get_location_id(): bigint {
  if (!_currentMap) return UNKNOWN_LOC_ID;
  const stack = capture_stack_frames();
  return _currentMap.intern_stack(stack);
}

File: libs/east-py/packages/east-py/east/expression/location.py
#
# Copyright (c) 2025 Elara AI Pty Ltd
# Licensed under the Business Source License 1.1. See LICENSE.md for details.
#
"""Authoring-frame source maps for python-built East functions (#626).

Every IR node a build constructs carries a ``loc_id``: an index into the
build's :class:`SourceMap`, whose entry is the stack of PYTHON frames that
were executing when the node was built — the lambda line that spelled the
expression, the ``East.function(...)`` call site, and the caller's frames
above it. A runtime error inside the compiled function resolves its node's
``loc_id`` back through that map, so ``EastError.location`` names the python
``file:line:column`` of the offending expression — on the east-py runner,
and on every other runner the function is exported to, because the map rides
the function value's beast2 encoding (the #476 wire shape reserves it).

This is the python twin of ``libs/east/src/location.ts``, contract for
contract: ``SourceMap`` interns stacks by content with entry 0 the empty
"no location" stack; :func:`capture_frames` walks the live python stack and
keeps only the AUTHOR's frames (East's own package, the standard library and
installed third-party packages are dropped, so a stack heads with user code
rather than with the builder internals); paths are relativized to a base
directory — the working directory unless :func:`set_location_base_path` says
otherwise — so serialized IR carries no machine-specific paths; and the map
is ambient: a build opens one with :func:`source_map_scope`, every node
built inside asks :func:`location_id` for its id, and a build nested inside
another (an ``East.function`` referenced from a body) shares the enclosing
map, exactly as ``ensure_source_map`` does in TS.

Capture is per NODE at build time and never per call: a compiled function
carries its map on the C side and resolves an error's ``loc_id`` only when
one is raised.
"""

from __future__ import annotations

import contextlib
import itertools
import os
import sys
import sysconfig
from collections.abc import Iterator
from typing import Any

__all__ = [
    "Location",
    "SourceMap",
    "UNKNOWN_LOC_ID",
    "capture_frames",
    "current_source_map",
    "location_id",
    "normalize_frame_path",
    "set_location_base_path",
    "source_map_scope",
]

#: One authoring frame: ``(filename, line, column)`` — line and column are
#: 1-based, matching TypeScript's ``Location`` (python reports 0-based
#: columns; they are shifted on capture).
Location = tuple[str, int, int]

#: The reserved "no / unknown location" id: ``resolve(0)`` is the empty stack.
UNKNOWN_LOC_ID = 0


class SourceMap:
    """Interned location stacks, indexed by ``loc_id``.

    Entry 0 is always the empty stack. Equal stacks intern to one id, so a
    node built from the same call site as an earlier one costs no entry —
    which is what keeps a map to roughly one entry per distinct expression.
    """

    __slots__ = ("_ids", "_stacks")

    def __init__(self) -> None:
        self._stacks: list[tuple[Location, ...]] = [()]
        self._ids: dict[tuple[Location, ...], int] = {}

    def intern_stack(self, stack: tuple[Location, ...]) -> int:
        """Intern a stack and return its stable id; the empty stack is 0."""
        if not stack:
            return UNKNOWN_LOC_ID
        loc_id = self._ids.get(stack)
        if loc_id is None:
            loc_id = len(self._stacks)
            self._stacks.append(stack)
            self._ids[stack] = loc_id
        return loc_id

    def resolve(self, loc_id: int) -> tuple[Location, ...]:
        """The stack ``loc_id`` names; empty for 0 or an unknown id."""
        if 0 <= loc_id < len(self._stacks):
            return self._stacks[loc_id]
        return ()

    def entries(self) -> list[tuple[Location, ...]]:
        """Every stack in id order (entry 0 is the empty stack) — the shape
        the wire formats and the C bridge consume."""
        return list(self._stacks)

    def __len__(self) -> int:
        """The number of entries, the reserved empty entry included."""
        return len(self._stacks)

    def __repr__(self) -> str:
        return f"<SourceMap {len(self._stacks) - 1} stack(s)>"


# ─── Frame path normalization ───────────────────────────────────────────────
#
# Frame paths are absolute and machine-specific. Baked into serialized IR
# they would leak the author's filesystem layout and make the IR differ from
# machine to machine, so every captured path is relativized to a base — the
# working directory by default, or what set_location_base_path() installs
# (deterministic fixtures) — and kept with forward slashes. Paths outside the
# base stay absolute; pseudo-files (``<stdin>``, ``<string>``) stay as they
# are. Never raises.

_explicit_base: str | None = None
_explicit_base_set = False
_auto_base: str | None = None
_auto_base_computed = False


def _clean(path: str) -> str:
    cleaned = path.replace("\\", "/")
    return cleaned[:-1] if len(cleaned) > 1 and cleaned.endswith("/") else cleaned


def _current_base() -> str | None:
    global _auto_base, _auto_base_computed
    if _explicit_base_set:
        return _explicit_base
    if not _auto_base_computed:
        _auto_base_computed = True
        try:
            _auto_base = _clean(os.getcwd())
        except OSError:
            _auto_base = None
    return _auto_base


def set_location_base_path(base: str | None) -> None:
    """Set the directory captured locations are relativized against.

    Pass a path to relativize against it (paths outside it stay absolute);
    pass ``None`` to reset to the automatic default, the working directory.
    The twin of TypeScript's ``setLocationBasePath``, for deterministic
    fixtures and reproducible exports.

    Args:
        base: The base directory, or ``None`` for the default.
    """
    global _explicit_base, _explicit_base_set
    if base is None:
        _explicit_base_set = False
        _explicit_base = None
    else:
        _explicit_base_set = True
        _explicit_base = _clean(os.path.abspath(base))
    _author_paths.clear()


def normalize_frame_path(raw: str) -> str:
    """A frame path in its portable form: forward slashes, relativized to
    the current base when it sits under it. Never raises."""
    try:
        cleaned = _clean(raw)
        base = _current_base()
        if base:
            if cleaned == base:
                return "."
            if cleaned.startswith(base + "/"):
                return cleaned[len(base) + 1:]
        return cleaned
    except Exception:  # pragma: no cover - a path nothing above expected
        return raw


# ─── Which frames are the author's ──────────────────────────────────────────
#
# East's own package builds the nodes, so its frames head every raw stack and
# would bake the builder's line numbers into every map; the standard library,
# installed (site-packages) third-party packages and the interpreter's script
# directory (the `pytest` / `east-py` entry scripts) are noise for the same
# reason. Everything else — the user's modules, a test file, a REPL — is the
# author. The decision is cached per code filename: a capture walks the whole
# python stack once per node, so each frame must cost one dict hit.

#: The ``east`` package directory (this file is east/expression/location.py).
_EAST_PACKAGE_DIR = os.path.dirname(os.path.dirname(os.path.abspath(__file__)))

_excluded_dirs: tuple[str, ...] | None = None

#: co_filename → its portable form, or None when the frame is not the author's.
_author_paths: dict[str, str | None] = {}
_MISS: Any = object()
_CACHE_LIMIT = 4096


def _excluded_directories() -> tuple[str, ...]:
    global _excluded_dirs
    if _excluded_dirs is None:
        dirs = {_EAST_PACKAGE_DIR}
        try:
            paths = sysconfig.get_paths()
            for key in ("stdlib", "platstdlib", "purelib", "platlib"):
                if paths.get(key):
                    dirs.add(os.path.abspath(paths[key]))
        except Exception:  # pragma: no cover - an exotic interpreter layout
            pass
        try:
            import site

            dirs.update(os.path.abspath(p) for p in site.getsitepackages())
            dirs.add(os.path.abspath(site.getusersitepackages()))
        except Exception:  # pragma: no cover - site is unavailable (embedded)
            pass
        if sys.executable:
            dirs.add(os.path.dirname(os.path.abspath(sys.executable)))
        _excluded_dirs = tuple(d.rstrip(os.sep) + os.sep for d in dirs)
    return _excluded_dirs


def _decide(filename: str) -> str | None:
    if filename.startswith("<"):
        # Pseudo-files: importlib's frozen modules are the interpreter's; a
        # REPL / exec'd string is the author's.
        return None if filename.startswith("<frozen ") else filename
    path = os.path.abspath(filename)
    for excluded in _excluded_directories():
        if path.startswith(excluded):
            return None
    return normalize_frame_path(path)


def _author_path(filename: str) -> str | None:
    hit = _author_paths.get(filename, _MISS)
    if hit is not _MISS:
        return hit
    if len(_author_paths) > _CACHE_LIMIT:
        _author_paths.clear()
    decision = _decide(filename)
    _author_paths[filename] = decision
    return decision


# ─── Positions ──────────────────────────────────────────────────────────────
#
# A frame's executing instruction has a (line, column) in its code object's
# position table (PEP 657, python 3.11+): the start of the expression being
# evaluated, which is the call the author wrote. Indexing the table is a walk,
# so positions memoize per (code object, instruction offset) — a call site
# builds many nodes, and an eager method captures the same lambda repeatedly.

_positions: dict[tuple[Any, int], tuple[int, int]] = {}


def _position(frame: Any) -> tuple[int, int]:
    code = frame.f_code
    lasti = frame.f_lasti
    key = (code, lasti)
    hit = _positions.get(key)
    if hit is not None:
        return hit
    line = frame.f_lineno
    column = 0
    try:
        index = lasti // 2
        pos = next(itertools.islice(code.co_positions(), index, index + 1), None)
    except Exception:  # pragma: no cover - a code object without positions
        pos = None
    if pos is not None:
        start_line, _end_line, start_col, _end_col = pos
        if start_line is not None:
            line = start_line
        if start_col is not None:
            column = start_col + 1
    if len(_positions) > _CACHE_LIMIT:
        _positions.clear()
    _positions[key] = (line, column)
    return line, column


def capture_frames() -> tuple[Location, ...]:
    """The author's frames of the current call, innermost first.

    East's own frames, the standard library and installed packages are
    dropped (see the module notes), paths are normalized, and columns are
    1-based. Empty when no author frame is on the stack.
    """
    frame = sys._getframe(1)
    out: list[Location] = []
    while frame is not None:
        path = _author_path(frame.f_code.co_filename)
        if path is not None:
            line, column = _position(frame)
            out.append((path, line, column))
        frame = frame.f_back
    return tuple(out)


# ─── The ambient map ────────────────────────────────────────────────────────

_current: SourceMap | None = None


@contextlib.contextmanager
def source_map_scope() -> Iterator[SourceMap]:
    """Run a build with a source map current, yielding that map.

    A fresh map when none is open; the ENCLOSING build's when one is — a
    function built inside another body shares its map, so the nested
    function's ids resolve wherever the enclosing one's do (TS
    ``ensure_source_map``). Re-entrant safe.
    """
    global _current
    if _current is not None:
        yield _current
        return
    _current = SourceMap()
    try:
        yield _current
    finally:
        _current = None


def current_source_map() -> SourceMap | None:
    """The map of the open build, or ``None`` outside any build."""
    return _current


def location_id() -> int:
    """The ``loc_id`` for the node being built right now.

    The interned authoring stack when a build is open; ``UNKNOWN_LOC_ID`` (0)
    when none is — nodes built outside any build (the internal helper
    functions, a type-only trace) carry no location, like TS's
    ``get_location_id`` outside a ``with_source_map`` scope.
    """
    current = _current
    if current is None:
        return UNKNOWN_LOC_ID
    return current.intern_stack(capture_frames())
